internal/cmd/dev.go

```diff
@@ -89,16 +98,390 @@ func (s *statement) UnmarshalJSON(data []byte) error {
+	err.Message = e.Error()
+	err.Code = &errSqliteUnknown
+	return err
+}
+
+type hranaCol struct {
+	Name string `json:"name"`
+}
+
+type hranaValue struct {
+	Type   string      `json:"type"`
+	Value  interface{} `json:"value,omitempty"`
+	Base64 string      `json:"base64,omitempty"`
+}
+
+type hranaNamedArg struct {
+	Name  string     `json:"name"`
+	Value hranaValue `json:"value"`
+}
+
+type hranaStmt struct {
+	Sql       *string          `json:"sql,omitempty"`
+	SqlId     *int32           `json:"sql_id,omitempty"`
+	Args      *[]hranaValue    `json:"args,omitempty"`
+	NamedArgs *[]hranaNamedArg `json:"named_args,omitempty"`
+	WantRows  *bool            `json:"want_rows"`
+}
+
+type hranaCond struct {
+	Type  string       `json:"type"`
+	Step  *int         `json:"step,omitempty"`
+	Cond  *hranaCond   `json:"cond,omitempty"`
+	Conds *[]hranaCond `json:"conds,omitempty"`
+}
+
+type hranaBatchStep struct {
+	Stmt *hranaStmt `json:"stmt,omitempty"`
+	Cond *hranaCond `json:"condition,omitempty"`
+}
+
+type hranaBatch struct {
+	Steps []hranaBatchStep `json:"steps"`
+}
+
+type hranaExecuteReq struct {
+	Stmt hranaStmt `json:"stmt"`
+}
+
+type hranaBatchReq struct {
+	Batch hranaBatch `json:"batch"`
+}
+
+type hranaStmtResult struct {
+	Cols     []hranaCol     `json:"cols"`
+	Rows     [][]hranaValue `json:"rows"`
+	Affected int32          `json:"affected_row_count"`
+	LastIR   *string        `json:"last_insert_rowid"`
+}
+
+type hranaSuccessResponse struct {
+	Result hranaStmtResult `json:"result"`
+}
+
+type hranaBatchResponse struct {
+	Results []*hranaStmtResult `json:"step_results"`
+	Errors  []*hranaError      `json:"step_errors"`
+}
+
+type hranaBatchSuccessResponse struct {
+	Result hranaBatchResponse `json:"result"`
+}
+
+func convertValue(val hranaValue) (interface{}, error) {
+	if val.Type == "blob" {
+		return []byte(val.Base64), nil
+	} else if val.Type == "integer" {
+		return strconv.Atoi(val.Value.(string))
+	} else {
+		return val.Value, nil
+	}
+
+}
+
+func executeHranaStatement(stmt hranaStmt, db *sql.DB) (*hranaStmtResult, *hranaError) {
+	var result hranaStmtResult
+
+	statement := *stmt.Sql
+	params := make([]interface{}, 0)
+
+	if stmt.Args != nil {
+		for _, narg := range *stmt.Args {
+			val, err := convertValue(narg)
+			if err != nil {
+				return nil, hranaErr(err)
+			}
+			params = append(params, val)
+		}
+	}
+
+	if stmt.NamedArgs != nil {
+		for _, narg := range *stmt.NamedArgs {
+			name := narg.Name
+			if strings.HasPrefix(name, ":") || strings.HasPrefix(name, "@") || strings.HasPrefix(name, "$") {
+				name = name[1:]
+			}
+
+			val, err := convertValue(narg.Value)
+			if err != nil {
+				return nil, hranaErr(err)
+			}
+
+			params = append(params, sql.Named(name, val))
+		}
+	}
+
+	if verboseFlag {
+		fmt.Printf("Executing %s, %s\n", internal.Emph(statement), internal.Emph(params))
+	}
+
+	result.Rows = make([][]hranaValue, 0)
+	result.Cols = make([]hranaCol, 0)
+
+	if shouldExec(statement, params...) {
```

Comment:
yeah, I don't like this, but I didn't find any good parser in go for sqlite. Without a parser, this will never work well. This is starting to change my mind on this approach. Maybe it is indeed better to just spawn sqld.